util/stmtsummary/evicted.go

```diff
@@ -190,3 +209,211 @@ func (seElement *stmtSummaryByDigestEvictedElement) toEvictedCountDatum() []type
 func (ssMap *stmtSummaryByDigestMap) ToEvictedCountDatum() [][]types.Datum {
 	return ssMap.other.ToEvictedCountDatum()
 }
+
+func (ssbde *stmtSummaryByDigestEvicted) toCurrentDatum() []types.Datum {
+	var seElement *stmtSummaryByDigestEvictedElement
+
+	ssbde.Lock()
+	if ssbde.history.Len() > 0 {
+		seElement = ssbde.history.Back().Value.(*stmtSummaryByDigestEvictedElement)
+	}
+	ssbde.Unlock()
+
+	if seElement == nil {
+		return nil
+	}
+
+	return seElement.toDatum()
+}
+
+func (ssbde *stmtSummaryByDigestEvicted) toHistoryDatum(historySize int) [][]types.Datum {
+	// Collect all history summaries to an array.
+	ssbde.Lock()
+	seElements := ssbde.collectHistorySummaries(historySize)
+	ssbde.Unlock()
+	rows := make([][]types.Datum, 0, len(seElements))
+
+	for _, seElement := range seElements {
+		rows = append(rows, seElement.toDatum())
+	}
+	return rows
+}
+
+func (ssbde *stmtSummaryByDigestEvicted) collectHistorySummaries(historySize int) []*stmtSummaryByDigestEvictedElement {
+	lst := make([]*stmtSummaryByDigestEvictedElement, 0, ssbde.history.Len())
+	for element := ssbde.history.Front(); element != nil && len(lst) < historySize; element = element.Next() {
+		seElement := element.Value.(*stmtSummaryByDigestEvictedElement)
+		lst = append(lst, seElement)
+	}
+	return lst
+}
+
+func (seElement *stmtSummaryByDigestEvictedElement) toDatum() []types.Datum {
+	return seElement.otherSummary.toDatum(new(stmtSummaryByDigest))
+}
+
+// addInfo adds information in addWith into addTo.
+func addInfo(addTo *stmtSummaryByDigestElement, addWith *stmtSummaryByDigestElement) {
+	addTo.Lock()
```

Comment:
Will `addWith` be modified at the same time?

Reply:
this struct is huge, so I passed its address to this function.

Reply:
No, my point is locking. Will it cause data race?

Reply:
`addWith` was locked outside the function, in `stmtSummaryByDigestEvicted.AddEvicted()` at line 90, so no lock needed here.